### PR TITLE
Add ncurses_layout option to customize the UI (buffer-list, grep summary…)

### PR DESCRIPTION
**Pull request by Delapouite**
Hi

## TL;DR

This PR removes `ncurses_status_on_top` option in favor of a more powerful and flexible `ncurses_layout`.

`ncurses_layout` is `main,status` by default. If you want the status on top behavior: `status,main`.

But it offers more potential. You can add extra `zones` (a better name can be discussed).
Let say you want to add a _buffer list_ on top of the screen.

1 - declare a regular string option, and name it like you desire, like `buffer_list`.
2 - add it to the `ncurses_layout` list. Example `buffer_list,main,status` to display the content of the `buffer_list` option in a new line on top of the screen.
3 - now it's up to a kak script to compute and update `buffer_list` when needed. It's value will be updated on screen. You can use faces.

That's it. No new commands involved, we just reuse and extend a feature pioneered by the `modelinefmt` option.

## UI paradigm

Kakoune already offers two ways to display extra bits of data on screen:
- `info`: only one at a time, no styles, get in the way if too big
- `menu`: only one at a time, interactive features

So having a system to just display `exta lines` of data like described in this PR provides an alternative persistent approach which is non intrusive and useful.

## Prior attempts

- https://github.com/mawww/kakoune/pull/1065:  this PR only focus on a buffer list. With `ncurses_layout` you can add as many _zones_ you need with any content you want.
- https://github.com/mawww/kakoune/wiki/Bar: this solution uses an external `bar` program to display things. It's hard to keep it displayed in the correct workspace, in the right position, always tied to the correct terminal window…
- [kakoune-electron](https://github.com/delapouite/kakoune-electron): using a complete foreign GUI system may have other advantages but it's a lot of work and rewriting for such a nice feature that can be implemented right in the current ncurses_ui and benefit every kakoune users out of the box.

## About the code

It's about 100ish new lines of code (including doc), with mainly plumbing stuffs for the different UIs (Remote, JSON…). Of course there's room to improve / fix things, it's mainly a proof of concept.

## Screenshot and example

image

```
declare-option str zone_buflist

define-command -hidden zone-buflist %{ %sh{
  list=''
  while read buf; do
    index=$(($index + 1))
    if [ "$buf" = "$kak_bufname" ]; then
      # markup specific to lemonbar
      list="$list{+r} $index $buf "
    else
      list="$list{Default} $index $buf "
    fi
  done <<< $(printf '%s\n' "$kak_buflist" | tr ':' '\n')
  echo "set-option global zone_buflist '$list'"
} }

hook global WinDisplay .* zone-buflist

set global ui_options ncurses_assistant=none:ncurses_layout=zone_buflist,main,status
```

### CODE REVIEW

**Comment by danr** [comment]
I like this, especially so since it generalises already existing functionality.

**Comment by Delapouite** [comment]
mawww's feedback on IRC:

- it would need to be a more general thing (having Client interpret ui_options seems wrong, having it read a ncurses_* value even more)
- we could isolate that in ncurses ui by using 2 options: the zones one, and have zones reference another entry in the ui_options map. `ui_options 'ncurses_zones=main,status,blah:blah=the text to have in blah zone'`. Then using hooks to update blah when buffer are opened/closed (if blah was used to list buffers). I believe using set -add on a  hash map option, such as ui_option should permit to just change the value of one entry 'set -add ui_options blah=...' will set/reset blah entry to whatever is given

**Comment by Delapouite** [comment]
The changes are now only limited to `ncurses_ui` and do not affect the client and other UIs anymore.

File: src/ncurses_ui.cc [hunk 1]
```diff
@@ -1061,9 +1082,31 @@ void NCursesUI::set_ui_options(const Options& options)
...
+        if (it != options.end())
+        {
+            auto layout = it->value | split<String>(',') | gather<Vector<String>>();
+            auto main = find(layout, "main");
+            auto status = find(layout, "status");
+            if (main == m_layout.end() or status == m_layout.end())
+                write_to_debug_buffer("main and status zones are mandatory in ncurses_layout");
+
+            m_layout = layout;
+            m_status_on_top = status < main;
```
Comment (mawww) [inline src/ncurses_ui.cc]: Do we still need m_status_on_top ? It looks like we should be able to get rid of it.

File: src/ncurses_ui.cc [hunk 2]
```diff
@@ -1061,9 +1082,31 @@ void NCursesUI::set_ui_options(const Options& options)
...
+        auto it = options.find("ncurses_layout"_sv);
+        if (it != options.end())
+        {
+            auto layout = it->value | split<String>(',') | gather<Vector<String>>();
+            auto main = find(layout, "main");
+            auto status = find(layout, "status");
+            if (main == m_layout.end() or status == m_layout.end())
+                write_to_debug_buffer("main and status zones are mandatory in ncurses_layout");
+
+            m_layout = layout;
```
Comment (mawww) [inline src/ncurses_ui.cc]: No need to copy the layout, use `m_layout = std::move(layout)` (and iterate on m_layout or move that line after the iteration).

File: src/ncurses_ui.cc [hunk 3]
```diff
@@ -1061,9 +1082,31 @@ void NCursesUI::set_ui_options(const Options& options)
...
-        auto it = options.find("ncurses_status_on_top"_sv);
-        m_status_on_top = it != options.end() and
-            (it->value == "yes" or it->value == "true");
+        auto it = options.find("ncurses_layout"_sv);
+        if (it != options.end())
+        {
+            auto layout = it->value | split<String>(',') | gather<Vector<String>>();
+            auto main = find(layout, "main");
+            auto status = find(layout, "status");
+            if (main == m_layout.end() or status == m_layout.end())
```
Comment (mawww) [inline src/ncurses_ui.cc]: if we get rid of the m_status_on_top assignment below, this can become `not contains(layout, "main") or not contains(layout, "status")`. Additionally, you are comparing to **m_**layout here, so that condition is going to be always false.

File: src/ncurses_ui.cc [hunk 4]
```diff
@@ -1106,4 +1149,15 @@ void NCursesUI::set_ui_options(const Options& options)
     }
 }
 
+int NCursesUI::get_zone_line (StringView zone_name)
```
Comment (mawww) [inline src/ncurses_ui.cc]: I dont like the use of `get_` here, should be `compute_` or just `zone_line`. Also, remove the space before `(`.

File: src/ncurses_ui.cc [hunk 5]
```diff
@@ -933,12 +954,13 @@ void NCursesUI::info_show(StringView title, StringView content,
...
 
-    const Rect rect = {m_status_on_top ? 1_line : 0_line, m_dimensions};
+    auto main_line_pos = get_zone_line("main");
+    const Rect rect = {{main_line_pos}, m_dimensions};
     InfoBox info_box;
     if (style == InfoStyle::Prompt)
     {
         info_box = make_info_box(m_info.title, m_info.content, m_dimensions.column, m_assistant);
-        anchor = DisplayCoord{m_status_on_top ? 0 : m_dimensions.line, m_dimensions.column-1};
+        anchor = DisplayCoord{m_status_on_top ? main_line_pos-1 : main_line_pos + m_dimensions.line, m_dimensions.column-1};
```
Comment (mawww) [inline src/ncurses_ui.cc]: I dont get why we use `m_status_on_top` here, the `get_zone_line` should give use the correct line offset.

**Comment by Clyybber** [comment]
Why was this closed?

**Comment by Delapouite** [comment]
Since faces were scoped, there is no easy way to rebase this PR.
Further efforts will continue on this branch which I intend to update soon: https://github.com/Delapouite/hakoune/tree/layout

**Comment by Jackojc** [comment]
Any update on this?

**Comment by sgraf812** [comment]
Well, I guess the answer is https://github.com/mawww/kakoune/wiki/Bar. But it would be neat to have all this logic wrapped in a plugin rather than having to copy and paste from a wiki page.

**Comment by Jackojc** [comment]
Wish there was a way to do it without external support of X or Wayland.
